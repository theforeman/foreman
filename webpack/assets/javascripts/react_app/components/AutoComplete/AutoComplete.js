import React from 'react';
import PropTypes from 'prop-types';
import { TypeAheadSelect } from 'patternfly-react';
import classNames from 'classnames';
import Immutable from 'seamless-immutable';
import { bindMethods, debounceMethods, noop } from '../../common/helpers';
import AutoCompleteMenu from './components/AutoCompleteMenu';
import AutoCompleteSearchButton from './components/AutoCompleteSearchButton';
import AutoCompleteError from './components/AutoCompleteError';
import AutoCompleteAux from './components/AutoCompleteAux';
import AutoCompleteFocusShortcut from './components/AutoCompleteFocusShortcut';
import { STATUS } from '../../constants';
import { TRIGGERS } from './AutoCompleteConstants';
import { KEYCODES } from '../../common/keyCodes';
import { translate as __ } from '../../common/I18n';
import './auto-complete.scss';

class AutoComplete extends React.Component {
  constructor(props) {
    super(props);
    bindMethods(this, [
      'handleClear',
      'handleInputChange',
      'handleResultsChange',
      'handleInputFocus',
      'getResults',
      'windowKeyPressHandler',
      'handleKeyDown',
    ]);
    this._typeahead = React.createRef();
    debounceMethods(this, 500, ['handleLoading']);
  }

  componentDidMount() {
    window.addEventListener('keypress', this.windowKeyPressHandler);
    const {
      controller,
      searchQuery,
      isDisabled,
      error,
      id,
      url,
      initialUpdate,
    } = this.props;

    initialUpdate({ searchQuery, controller, id, isDisabled, error, url });
  }

  componentDidUpdate(prevProps) {
    this.handleLoading();
    const { searchQuery, trigger } = this.props;
    const { RESET, CONTROLLER_CHANGED } = TRIGGERS;
    if (trigger === RESET || trigger === CONTROLLER_CHANGED) {
      this.handleClear();
    }
    if (prevProps.searchQuery !== searchQuery) {
      const typeahead = this._typeahead && this._typeahead.current;
      typeahead && typeahead.setState({ text: searchQuery });
    }
  }

  windowKeyPressHandler(e) {
    const { useKeyShortcuts, handleSearch } = this.props;
    const instance = this._typeahead.current.getInstance();
    const { ENTER, FWD_SLASH, BACK_SLASH } = KEYCODES;
    const { tagName } = e.target;
    const didEventCameFromInput = tagName === 'INPUT' || tagName === 'TEXTAREA';

    /**
     Disable this functionality if the event came from an input,
     or if the 'useKeyShortcuts' is falsy.
    */
    if (didEventCameFromInput || !useKeyShortcuts) {
      return;
    }

    switch (e.charCode) {
      case ENTER: {
        handleSearch();
        break;
      }
      case FWD_SLASH:
      case BACK_SLASH: {
        const {
          focus,
          state: { showMenu },
        } = instance;
        const isMenuHidden = !showMenu;
        if (isMenuHidden) {
          e.preventDefault();
          focus();
        }
        break;
      }
      default: {
        break;
      }
    }
  }

  getResults(searchQuery, trigger, id) {
    const { getResults, controller, url } = this.props;
    getResults({
      url,
      searchQuery,
      controller,
      trigger,
      id,
    });
  }

  handleInputFocus({ target: { value } }) {
    const { id, results } = this.props;
    if (results.length === 0) {
      this.getResults(value, TRIGGERS.INPUT_FOCUS, id);
    }
  }

  handleInputChange(searchQuery) {
    const { id } = this.props;
    this.getResults(searchQuery, TRIGGERS.INPUT_CHANGE, id);
  }

  // Gets the first result from an array of selected results.
  handleResultsChange({ 0: result }) {
    const { id } = this.props;
    if (!result) {
      return;
    }
    this.getResults(result, TRIGGERS.ITEM_SELECT, id);
    /**
     *  HACK: I had no choice but to call to an inner function,
     * due to lack of design in react-bootstrap-typeahead.
     */
    this._typeahead.current.getInstance()._showMenu();
  }

  handleKeyDown({ keyCode }) {
    const instance = this._typeahead.current.getInstance();
    switch (keyCode) {
      case KEYCODES.ENTER: {
        if (!instance.state.activeItem) {
          this.props.handleSearch();
        }
        break;
      }
      case KEYCODES.ESC: {
        instance.blur();
        break;
      }
      default: {
        break;
      }
    }
  }

  handleClear() {
    const { id } = this.props;
    this._typeahead.current.getInstance().clear();
    this.getResults('', TRIGGERS.INPUT_CLEAR, id);
  }

  handleLoading() {
    const { status } = this.props;
    const typeahead = this._typeahead && this._typeahead.current;
    const isLoading = status === STATUS.PENDING;
    typeahead && typeahead.setState({ isLoading });
  }

  componentWillUnmount() {
    window.removeEventListener('keypress', this.windowKeyPressHandler);
    const { resetData, controller, id } = this.props;
    resetData(controller, id);
  }

  render() {
    const {
      id,
      error,
      searchQuery,
      inputProps,
      placeholder,
      results,
      useKeyShortcuts,
      isDisabled,
    } = this.props;
    /** Using a 3rd party library (react-bootstrap-typeahead) that expects a mutable array. */
    const options = Immutable.isImmutable(results)
      ? results.asMutable()
      : results;

    return (
      <div className="foreman-autocomplete">
        <TypeAheadSelect
          id={id}
          ref={this._typeahead}
          defaultInputValue={searchQuery}
          options={options}
          onInputChange={this.handleInputChange}
          onChange={this.handleResultsChange}
          onFocus={this.handleInputFocus}
          onKeyDown={this.handleKeyDown}
          placeholder={__(placeholder)}
          disabled={isDisabled}
          renderMenu={(r, menuProps) => (
            <AutoCompleteMenu {...{ results: r, menuProps }} />
          )}
          inputProps={{
            className: classNames(
              'search-input',
              useKeyShortcuts ? 'use-shortcuts' : ''
            ),
            spellCheck: 'false',
            'data-autocomplete-id': id,
            autoComplete: 'off',
            ...inputProps,
          }}
        />
        <AutoCompleteAux onClear={this.handleClear} />
        <AutoCompleteFocusShortcut useKeyShortcuts={useKeyShortcuts} />
        <AutoCompleteError error={error} />
      </div>
    );
  }
}

AutoComplete.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  url: PropTypes.string.isRequired,
  results: PropTypes.array,
  searchQuery: PropTypes.string,
  inputProps: PropTypes.object,
  status: PropTypes.string,
  error: PropTypes.string,
  initialError: PropTypes.string,
  controller: PropTypes.string,
  handleSearch: PropTypes.func,
  getResults: PropTypes.func,
  resetData: PropTypes.func,
  initialUpdate: PropTypes.func,
  useKeyShortcuts: PropTypes.bool,
  placeholder: PropTypes.string,
  isDisabled: PropTypes.bool,
  trigger: PropTypes.string,
};

AutoComplete.defaultProps = {
  results: [],
  searchQuery: '',
  inputProps: {},
  status: null,
  error: null,
  initialError: null,
  controller: null,
  handleSearch: noop,
  getResults: noop,
  resetData: noop,
  initialUpdate: noop,
  useKeyShortcuts: false,
  placeholder: 'Filter ...',
  isDisabled: false,
  trigger: null,
};

AutoComplete.SearchButton = AutoCompleteSearchButton;
AutoComplete.Error = AutoCompleteError;

export default AutoComplete;
