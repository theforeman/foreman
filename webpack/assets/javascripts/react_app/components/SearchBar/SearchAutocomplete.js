import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { Menu, MenuContent, Popper, SearchInput } from '@patternfly/react-core';
import classNames from 'classnames';
import { AutoCompleteMenu } from './AutoCompleteMenu';
import { translate as __ } from '../../common/I18n';
import './SearchBar.scss';

export const SearchAutocomplete = ({
  results,
  onSearchChange,
  value,
  onSearch,
  disabled,
  error,
  name,
}) => {
  const [isAutocompleteOpen, setIsAutocompleteOpen] = useState(false);
  const searchWrapperRef = useRef(null);
  const searchInputRef = useRef(null);
  const autocompleteRef = useRef(null);

  const _onSearch = searchValue => {
    setIsAutocompleteOpen(false);
    onSearch && onSearch(searchValue);
  };
  const onClear = () => {
    onSearchChange('');
    _onSearch('');
  };

  const onChange = (newValue, e) => {
    if (searchInputRef?.current?.contains(document.activeElement)) {
      setIsAutocompleteOpen(true);
      onSearchChange(newValue);
    }
  };

  const onSelect = (e, itemId) => {
    e.stopPropagation();
    if (itemId[itemId.length - 1] !== ' ') {
      itemId = `${itemId} `;
    }
    if (itemId[0] === ' ') {
      itemId = itemId.slice(1);
    }
    onSearchChange(itemId);
    searchInputRef.current.focus();
  };

  useEffect(() => {
    const handleMenuKeys = event => {
      // keyboard shortcut to focus the search, will not focus if the key is typed into an input
      if (
        event.key === '/' &&
        event.target.tagName !== 'INPUT' &&
        event.target.tagName !== 'TEXTAREA'
      ) {
        event.preventDefault();
        searchInputRef.current.focus();
      }
      // if the autocomplete is open and the browser focus is on the search input,
      else if (isAutocompleteOpen && searchInputRef?.current === event.target) {
        // the escape key closes the autocomplete menu and keeps the focus on the search input.
        if (event.key === 'Escape') {
          setIsAutocompleteOpen(false);
          searchInputRef.current.focus();
          // the up and down arrow keys move browser focus into the autocomplete menu
        } else if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
          const firstElement = autocompleteRef.current.querySelector(
            'li > button:not(:disabled)'
          );
          firstElement && firstElement.focus();
          event.preventDefault(); // by default, the up and down arrow keys scroll the window
        } else if (
          autocompleteRef?.current?.contains(event.target) &&
          event.key === 'Tab'
        ) {
          event.preventDefault();
          setIsAutocompleteOpen(false);
          searchInputRef.current.focus();
        }
      }
    };
    // The autocomplete menu should close if the user clicks outside the menu.
    const handleClickOutside = event => {
      if (
        isAutocompleteOpen &&
        searchWrapperRef &&
        searchWrapperRef.current &&
        !searchWrapperRef.current.contains(event.target)
      ) {
        setIsAutocompleteOpen(false);
      }
    };
    window.addEventListener('keydown', handleMenuKeys);
    window.addEventListener('click', handleClickOutside);

    return () => {
      window.removeEventListener('keydown', handleMenuKeys);
      window.removeEventListener('click', handleClickOutside);
    };
  }, [isAutocompleteOpen]);
  const searchInput = (
    <div
      ref={searchWrapperRef}
      onClick={e => {
        if (e.target.type !== 'submit') setIsAutocompleteOpen(true);
      }}
      className={classNames('autocomplete-search', {
        'disabled-autocomplete-search': disabled,
      })}
    >
      <SearchInput
        ref={searchInputRef}
        value={value}
        onChange={(_e, newValue) => {
          onChange(newValue);
        }}
        onClear={onClear}
        onSearch={(event, searchValue) => {
          if (onSearch) {
            onSearch(searchValue);
          } else {
            _onSearch(searchValue);
          }
        }}
        isDisabled={disabled}
        placeholder={__('Search')}
        resetButtonLabel={__('Reset search')}
        name={name}
      />
    </div>
  );

  const autocomplete = (
    <Menu
      ouiaId="search-autocomplete-menu"
      ref={autocompleteRef}
      onSelect={error ? null : onSelect}
    >
      {!disabled && (
        <MenuContent>
          <AutoCompleteMenu results={results} error={error} />
        </MenuContent>
      )}
    </Menu>
  );

  return (
    <Popper
      trigger={searchInput}
      popper={autocomplete}
      isVisible={isAutocompleteOpen}
      enableFlip={false}
    />
  );
};

SearchAutocomplete.propTypes = {
  results: PropTypes.array.isRequired,
  onSearchChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  onSearch: PropTypes.func,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  name: PropTypes.string,
};

SearchAutocomplete.defaultProps = {
  onSearch: null,
  disabled: false,
  error: null,
  name: null,
};
