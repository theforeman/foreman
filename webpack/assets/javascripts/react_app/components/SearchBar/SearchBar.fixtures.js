export const SearchBarProps = {
  data: {
    autocomplete: {
      searchQuery: null,
      url: 'model/auto_complete_search',
    },
    bookmarks: {
      url: '/api/bookmarks',
      canCreate: true,
      documentationUrl: '/doc/url',
    },
    controller: 'models',
  },
};

export const mockResults = [
  { label: 'name', category: '' },
  { label: 'info', category: '' },
  { label: ' not', category: 'Operators' },
  { label: ' has', category: 'Operators' },
];
