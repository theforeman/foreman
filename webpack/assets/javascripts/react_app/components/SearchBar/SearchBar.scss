@import '~@theforeman/vendor/scss/variables';

.autocomplete-search {
  width: 100%;
  &.disabled-autocomplete-search .pf-c-input-group{
    color: var(--pf-global--disabled-color--100);
    background-color: var(--pf-global--disabled-color--300);
    cursor: not-allowed;
    border-color: var(--pf-global--disabled-color--300);
  }
}

.foreman-search-bar {
  display: flex;
  .pf-c-dropdown.pf-m-align-right {
    width: unset;
  }
  .pf-c-dropdown { 
    width: unset; // just in case Katello overrides it
  }
}
