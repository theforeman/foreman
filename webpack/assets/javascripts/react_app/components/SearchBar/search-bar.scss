@import "../AutoComplete/auto-complete.scss";
@import '~bootstrap-sass/assets/stylesheets/bootstrap/_variables';

.search-bar {
  .autocomplete-aux {
    right: 108px;
    // Keeps in position when screen size is xs and search button shrinks.
    @media screen and (max-width: $screen-sm-min) {
      right: 65px;
    }
  }

  .autocomplete-search-btn-text {
    display: inline;
  }

  .search-input.use-shortcuts {
    padding-left: 21px;
  }
}
