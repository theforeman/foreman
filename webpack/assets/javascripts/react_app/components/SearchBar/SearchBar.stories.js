import React from 'react';
import { Provider } from 'react-redux';
import { storiesOf } from '@storybook/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import URI from 'urijs';
import Store from '../../redux';
import SearchBar from './index';
import { SearchBarProps, APImock } from './SearchBar.fixtures';
import Story from '../../../../../stories/components/Story';

const someAutoCompletePath = /^models\/auto_complete_search.*/;
const mock = new MockAdapter(axios);
mock.onGet(someAutoCompletePath).reply(({ url }) => {
  const query = URI.decodeQuery(url.split('search=')[1]);
  const results = APImock[query];
  return [200, results || []];
});

storiesOf('Components/SearchBar', module).add(
  'Search Bar with mocked data',
  () => (
    <Provider store={Store}>
      <Story narrow>
        <h4>Try typing something like: &quot;name = &quot;</h4>
        <SearchBar data={SearchBarProps.data} />
      </Story>
    </Provider>
  )
);
