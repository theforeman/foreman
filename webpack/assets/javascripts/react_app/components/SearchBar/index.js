import { connect } from 'react-redux';
import SearchBar from './SearchBar';
import { selectAutocompleteSearchQuery } from '../AutoComplete/AutoCompleteSelectors';

const mapStateToProps = state => ({
  searchQuery: selectAutocompleteSearchQuery(state),
});

export default connect(mapStateToProps)(SearchBar);
