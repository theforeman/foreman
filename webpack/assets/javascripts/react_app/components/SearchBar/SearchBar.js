import React from 'react';
import isEmpty from 'lodash/isEmpty';
import PropTypes from 'prop-types';
import URI from 'urijs';
import AutoComplete from '../AutoComplete';
import Bookmarks from '../bookmarks';
import './search-bar.scss';

const handleSearch = searchQuery => {
  const uri = new URI(window.location.href);
  const data = { ...uri.query(true), search: searchQuery.trim(), page: 1 };
  uri.query(URI.buildQuery(data, true));
  window.Turbolinks.visit(uri.toString());
};

const SearchBar = ({
  searchQuery,
  data: { autocomplete, controller, bookmarks },
}) => {
  const bookmarksComponent = !isEmpty(bookmarks) ? (
    <Bookmarks data={{ ...bookmarks, controller, searchQuery }} />
  ) : null;
  return (
    <div className="search-bar input-group">
      <AutoComplete
        handleSearch={() => handleSearch(searchQuery)}
        initialQuery={autocomplete.searchQuery || ''}
        useKeyShortcuts={autocomplete.useKeyShortcuts}
        url={autocomplete.url}
        controller={controller}
      />
      <div className="input-group-btn">
        <AutoComplete.SearchButton onClick={() => handleSearch(searchQuery)} />
        {bookmarksComponent}
      </div>
    </div>
  );
};

SearchBar.propTypes = {
  searchQuery: PropTypes.string,
  data: PropTypes.shape({
    autocomplete: PropTypes.shape({
      results: PropTypes.array,
      searchQuery: PropTypes.string,
      url: PropTypes.string,
      useKeyShortcuts: PropTypes.bool,
    }),
    controller: PropTypes.string,
    bookmarks: PropTypes.shape({ ...Bookmarks.propTypes }),
  }),
};

SearchBar.defaultProps = {
  searchQuery: '',
  data: {
    autocomplete: {
      results: [],
      searchQuery: null,
      url: null,
      useKeyShortcuts: true,
    },
    controller: null,
    bookmarks: {},
  },
};

export default SearchBar;
