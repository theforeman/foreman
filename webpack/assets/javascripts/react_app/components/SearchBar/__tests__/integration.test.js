import React from 'react';
import API from '../../../API';
import IntegrationTestHelper from '../../../common/IntegrationTestHelper';
import { SearchBarProps } from '../SearchBar.fixtures';
import SearchBar from '../index';
import { reducers } from '../../AutoComplete';
import bookmarksReducer from '../../../redux/reducers/bookmarks';
import { KEYCODES } from '../../AutoComplete/AutoCompleteConstants';

jest.mock('../../../API');
jest.mock('lodash/debounce', () => jest.fn(fn => fn));
global.Turbolinks = {
  visit: jest.fn(),
};

const combinedReducers = { ...reducers, bookmarks: bookmarksReducer };

describe('SearchBar integration test', () => {
  it('should flow', async () => {
    const label = 'result';
    API.get.mockImplementation(async () => ({
      data: [{ label, category: '' }],
    }));
    const integrationTestHelper = new IntegrationTestHelper(combinedReducers);
    const wrapper = integrationTestHelper.mount(
      <SearchBar {...SearchBarProps} />
    );
    const autocomplete = wrapper.find('AutoComplete').instance();
    autocomplete.handleInputChange(label);
    // trigger search button click.
    wrapper
      .find('.autocomplete-search-btn')
      .first()
      .simulate('click');
    // expect it to call Turbolinks.
    expect(global.Turbolinks.visit.mock.calls).toHaveLength(1);
    const event = new KeyboardEvent('keypress', { charCode: KEYCODES.ENTER });
    global.dispatchEvent(event);
    expect(global.Turbolinks.visit.mock.calls).toHaveLength(2);
    // bookmark this page:
    // click on bookmark button
    wrapper
      .find('button[title="Bookmarks"]')
      .first()
      .simulate('click');
    integrationTestHelper.takeStoreAndLastActionSnapshot(
      'bookmarks button click'
    );
    // dropdown should open, lets click on 'Bookmark this page'
    wrapper
      .find('a[id="newBookmark"]')
      .first()
      .simulate('click');
    integrationTestHelper.takeStoreAndLastActionSnapshot(
      'in bookmarks dropdown: click on "bookmark this page"'
    );
    // modal should open, lets check its query value
    expect(
      wrapper
        .find('ConnectedField')
        .at(1)
        .props().initial
    ).toBe(label);
  });
});
