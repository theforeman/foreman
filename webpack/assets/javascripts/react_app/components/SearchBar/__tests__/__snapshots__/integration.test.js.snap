// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar integration test should flow: bookmarks button click 1`] = `
Object {
  "action": Array [
    Array [
      Object {
        "key": "BOOKMARKS",
        "payload": Object {
          "controller": "models",
        },
        "type": "API_GET",
        "url": "/api/bookmarks?search=controller%3Dmodels&per_page=100",
      },
    ],
  ],
  "state": Object {
    "autocomplete": Object {
      "some-id": Object {
        "controller": "models",
        "error": null,
        "isErrorVisible": false,
        "searchQuery": "result",
        "status": "PENDING",
        "trigger": "INPUT_CHANGE",
      },
    },
    "bookmarks": Object {
      "models": Object {
        "errors": null,
        "results": Array [],
        "status": "PENDING",
      },
      "showModal": false,
    },
  },
}
`;

exports[`SearchBar integration test should flow: in bookmarks dropdown: click on "bookmark this page" 1`] = `
Object {
  "action": Array [
    Array [
      Object {
        "payload": Object {
          "query": "result",
        },
        "type": "BOOKMARKS_MODAL_OPENED",
      },
    ],
  ],
  "state": Object {
    "autocomplete": Object {
      "some-id": Object {
        "controller": "models",
        "error": null,
        "isErrorVisible": false,
        "searchQuery": "result",
        "status": "PENDING",
        "trigger": "INPUT_CHANGE",
      },
    },
    "bookmarks": Object {
      "currentQuery": "result",
      "models": Object {
        "errors": null,
        "results": Array [],
        "status": "PENDING",
      },
      "showModal": true,
    },
  },
}
`;
